### DISCUSSION TITLE
[@types/react] Version 18.2.69 and Above: Void Problem with useRef Callback

### DISCUSSION THREAD

**Discussion by GreattitJY** [DefinitelyTyped/DefinitelyTyped#69416] [Issues with a @types package] [answered]
Since updating to @types/react version 18.2.69 and above, there's been an issue with the useRef hook where passing an element (el) into the ref callback function results in a void problem. This wasn't an issue in versions 18.2.68 and below. To resolve this, it appears necessary to modify all existing code from:

`ref={(el) => (listItemRef.current[i] = el)}`

to:

`ref={(el) => {(listItemRef.current[i] = el)}}`.

Encountering this problem while using '^' to ensure compatibility with versions has been confusing. Shouldn't updates be backward compatible? 

```tsx 
import { forwardRef } from "react";

interface ListItemProps extends React.HTMLAttributes<HTMLLIElement> {}

const ListItem = forwardRef<HTMLLIElement, ListItemProps>(({}, ref) => {
  return <li ref={ref}></li>;
});

ListItem.displayName = "ListItem";

export { ListItem };
```

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/103429329/b7b87e3f-9e9b-4d13-816f-febf231141e0)
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/103429329/ac7d357e-a5ed-4e6f-a0db-c58ab46d51b2)

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
You're probably getting types for Canary. But that's not necessarily relevant here since that will be the behavior going forward.

Unfortunately, TypeScript will allow any returntype of functions that can be assigned to `() => void`. We should've always flagged the original code. 

The reason that only surfaces in Canary types and will surface in React 19 types is that ref callbacks can now return a cleanup function. So we're going from `() => void` to `() => (void | (() => void))`. Now TypeScript will reject any return type that's not `void` or `() => void`.

To help migration, you can use the [`no-implicit-ref-callback-return` codemod](https://github.com/eps1lon/types-react-codemod#no-implicit-ref-callback-return) like so: `npx types-react-codemod no-implicit-ref-callback-return src/`
